segments_tree.go

```diff
@@ -10,13 +10,20 @@ const SegmentSeparator = "\n"
 type segmentsTree struct {
 	root bool
 
+	// "nodes" stores a map from a segment to it's childrens
+	// in an hierarchial data like JSON, a node can be Object or Array.
+	// for example, in this path "context\nuser\nid", both "context" and "user" will be nodes.
 	nodes map[string]*segmentsTree
 
-	// fields map segment to full path
-	// will be present only on the leafs
+	// "fields" is storing leaf fields here mapped to their full paths
```

Comment:
I suggest we find a different name than "fields" because `Field` is already a well-known widely-used Quamina concept.  I notice your comment says "…is storing leaf nodes" so maybe "leaves" would be a better name?  Or something else.  

Hmm, thinking out loud, the leaf name is the key and the value is (in Quamina terms) `Field.Path`. So maybe this could be called `leafPaths` or something?

Reply:
Yeah, the name could be better here.. I couldn't figure out which name to call it.

If we are talking about JSON-land it should be named "properties", if we are talking about Quamina's land it should be called "segments" but it's the last segment of path.

When we will call it "leaves" it will signal that this data is stored only on leaf nodes (in our tree), but this data can be actually stored at any level, think about the following paths (using dot as separator):
1. `context.user_id` - will be stored on the `context` node (under root)
2. `context.user.id` - will be stored on the `user` node (under `context`)
3. `user_id` - will be stored on the root node

`leafPaths` is ok, but don't you think it will be confusing since we are storing segments as keys?